### PR TITLE
fix(gms): filter out runs of a dataJob without any run-events

### PR DESCRIPTION
**Pull request by ksrinath**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by david-leifker — CHANGES_REQUESTED** [review]
None of this code is required to add a hasXXX field and this implementation is not correctly handling the ES index mapping.

This feature should be implemented by adding a @Searchable annotation to the entity on the timeseries field in the pdl model. This is done with an annotation called `hasValuesFieldName` and will both update the ES mapping and populate the boolean without any additional code beyond the addition to the annotation.

File: metadata-models/src/main/pegasus/com/linkedin/dataprocess/DataProcessInstanceRunEvent.pdl [hunk 1]
```diff
@@ -15,6 +15,10 @@ import com.linkedin.common.Urn
 record DataProcessInstanceRunEvent includes TimeseriesAspectBase, ExternalReference {
 
   @TimeseriesField = {}
+  @Searchable = {
+    "fieldType": "TEXT",
```
Comment (david-leifker) [inline metadata-models/src/main/pegasus/com/linkedin/dataprocess/DataProcessInstanceRunEvent.pdl:19]: This should either be removed or set to KEYWORD. The reason is that the value is of type enum which would default to KEYWORD.

**Review by david-leifker — APPROVED** [review]

**Review by david-leifker — CHANGES_REQUESTED** [review]
The upgrade job looks really good! 

1. Add a `delayMs` to the primary `while` loop. This should be set to a default of 1000 millis. The reason is that we want to throttle a little as the aggregation query can put load on the ES cluster and allows us to increase the delay if it triggers alerts on the cluster.
2. The aggregation query should specify a time window start/stop on the timeseries aspect index. We've seen negative performance impacts from querying the entire timeseries indices (A customer with large timeseries would be Robinhood for example). I would expect the query to iterate on buckets of time. For example, a 1 day window. - Bonus points if the window is fully configurable, but a good conservative default would be to aggregate timeseries data 1 day at a time and maybe assume that it processes the last 90d.

**Review by david-leifker — CHANGES_REQUESTED** [review]
The aggregation query will only return `batchSize` (top N) urns from the timeseries index. How does this code not just update the `batchSize` most frequent entities?

**Comment by ksrinath** [comment]
> The aggregation query will only return `batchSize` (top N) urns from the timeseries index. How does this code not just update the `batchSize` most frequent entities?

The `afterKey` is updated in the outer-infinite-while loop, so that all the distinct-urns should get read by the time we break out of the while-loop. Am I missing something here?

**Comment by ksrinath** [comment]
> 1. Add a `delayMs` to the primary `while` loop. This should be set to a default of 1000 millis. The reason is that we want to throttle a little as the aggregation query can put load on the ES cluster and allows us to increase the delay if it triggers alerts on the cluster.

Updated code for this; can you please review.

**Comment by david-leifker** [comment]
> The aggregation query will only return batchSize (top N) urns from the timeseries index. How does this code not just update the batchSize most frequent entities?
> 
> The afterKey is updated in the outer-indefinite-while loop, so that all the distinct-urns should get read by the time we break out of the while-loop. Am I missing something here?

RIght, it *is* using the *Composite* aggregator here. It should work.

File: metadata-service/configuration/src/main/resources/application.yaml [hunk 2]
```diff
@@ -370,7 +370,12 @@ systemUpdate:
     batchSize: ${SYSTEM_UPDATE_SCHEMA_FIELDS_DOC_IDS_BATCH_SIZE:500}
     delayMs: ${SYSTEM_UPDATE_SCHEMA_FIELDS_DOC_IDS_DELAY_MS:5000}
     limit: ${SYSTEM_UPDATE_SCHEMA_FIELDS_DOC_IDS_LIMIT:0}
-
+  processInstanceHasRunEvents:
```
Comment (david-leifker) [inline metadata-service/configuration/src/main/resources/application.yaml:373]: Missing `delayMs` maybe the code isn't pushed?

**Comment by david-leifker** [comment]
@ksrinath - I am not seeing `delayMs` - missing changes?

File: metadata-service/configuration/src/main/resources/application.yaml (hunk 2 above)
Comment (ksrinath) [reply metadata-service/configuration/src/main/resources/application.yaml:373]: Yes, apologies - pushed now.

**Comment by ksrinath** [comment]
> 2. The aggregation query should specify a time window start/stop on the timeseries aspect index. We've seen negative performance impacts from querying the entire timeseries indices (A customer with large timeseries would be Robinhood for example). I would expect the query to iterate on buckets of time. For example, a 1 day window. - Bonus points if the window is fully configurable, but a good conservative default would be to aggregate timeseries data 1 day at a time and maybe assume that it processes the last 90d.

Can you please review the updated code.

**Review by david-leifker — APPROVED** [review]
LGTM! Great job!

**Pull request merged**